File: src/agent/telegram.ts
import {
  ITelegramConfig,
  IAuthSession,
  IAuthenticator,
  IAuthSessionEvt,
  IControlCenter
} from "../api";
import Alarm from "../alarm";
import Camera from "./camera";
import Telegraf, { ContextMessageUpdate } from "telegraf";
import * as tt from "telegraf/typings/telegram-types.d";
import * as log4js from "log4js";
import { instance as authSessionManager, AuthStates } from "../authSessionManager";
import * as util from "../util";
import * as moment from "moment";

const Markup = require("telegraf/markup");

const logger = log4js.getLogger("telegramBot");
const cmdRegex = /^\/([^\s]+)\s*(.*)$/; // command example /photo <params>
const AUTH_MSG = "Intrusion detected. What is your password ?";
export class TelegramAgent implements IAuthenticator, IControlCenter {
  name = "telegram";
  delay: moment.Duration;
  private bot: Telegraf<ContextMessageUpdate>;
  private sessionId: string; // running session
  private answerExpected: string;
  private lastMessageSentTime: number;
  private cmdHandlers: {
    [id: string]: (...args: any[]) => Promise<any>;
  } = {};
  private lastMsgIdProcessed: number;
  private chatId: string;

  constructor(public telegramCfg: ITelegramConfig, public alarm: Alarm, public camera: Camera) {
    this.bot = new Telegraf(telegramCfg.botToken);

    for (const meth of [
      this.onDisable,
      this.onDisarm,
      this.onEnable,
      this.onPhoto,
      this.onStatus,
      this.onStart,
      this.onCam,
      this.onCamStatus
    ]) {
      const methName = util.getMemberFunctionName(meth);
      const evtName = methName.substring(2).toLowerCase();
      this.cmdHandlers[evtName] = meth.bind(this);
      logger.debug("Added command %s with handler method %s", evtName, methName);
    }
    this.bot.on("message", this.onMessage.bind(this));
  }

  async start() {
    logger.debug("Starting telegram agent");

    this.bot.startPolling(30);
    logger.debug("telegram agent started");
    return await Promise.resolve();
  }

  async onMessage(ctx: ContextMessageUpdate): Promise<any> {
    if (!ctx.from || "" + ctx.from.id !== this.telegramCfg.userId) {
      logger.debug("Unidentifed message emitter, got %j", ctx.from);
      return;
    } else {
      logger.debug("Got message %j", ctx.message);
    }
    if (this.lastMsgIdProcessed === ctx.message.message_id) {
      logger.debug("Already processed, dropping");
      return;
    }
    this.chatId = (ctx.chat && ctx.chat.id) + "";
    this.lastMsgIdProcessed = ctx.message.message_id;
    const cmdMatch = cmdRegex.exec(ctx.message.text);
    let handler;

    if (cmdMatch) {
      // process control center command
      const [, cmd, cmdArgs] = cmdMatch;

      handler = this.cmdHandlers[cmd.toLowerCase()];
      if (handler) {
        try {
          await handler(cmdArgs);
        } catch (err) {
          logger.error("Failed processing cmd [%s]", cmd, err);
        }
      }
    } else if (this.answerExpected) {
      // process ask reply
      const eventType = this.answerExpected;
      delete this.answerExpected;
      handler = this.cmdHandlers[eventType];
      if (handler) {
        try {
          await handler(ctx.message.text);
        } catch (err) {
          logger.error("Failed processing chat event [%s]", eventType, err);
        }
      }
    } else if (this.sessionId) {
      // process auth session flow
      const session = authSessionManager.getSession(this.sessionId);
      if (session) {
        try {
          await this.onAuthSessionMessage(session, ctx.message.text);
        } catch (err) {
          session.reportFailure(err, "telegram");
        }
      }
    }
  }

  private async sendMessage(text: string, opts?: tt.ExtraReplyMessage) {
    this.lastMessageSentTime = Date.now();
    if (this.chatId) {
      return await this.bot.telegram.sendMessage(this.chatId, "[Alarm] " + text, opts);
    } else {
      throw new Error("No chat id");
    }
  }

  async onDisable() {
    try {
      const isAuthSessionRunning = !!this.sessionId;
      await this.alarm.disable(this.name);
      if (!isAuthSessionRunning) {
        return await this.onStatus();
      }
    } catch (err) {
      logger.error("could not disable alarm", err);
      return await this.sendMessage("Failed disabling alarm");
    }
  }

  async onEnable() {
    try {
      await this.alarm.enable();
      return await this.onStatus();
    } catch (err) {
      logger.error("could not enable alarm", err);
      return await this.sendMessage("failed enabling alarm");
    }
  }

  async onStatus(msg?: string) {
    let txtMsg;
    if (msg) {
      txtMsg = msg + "\n";
    }
    // TODO: remove check for testing (avoiding to instantiate an alarm)
    if (this.alarm) {
      txtMsg = "status:" + this.alarm.state.name;
      if (this.alarm.state === this.alarm.STATES.DISARMED) {
        txtMsg += "\n will be enabled " + this.alarm.nextEnableDate;
      }
    }
    return await this.sendMessage(txtMsg);
  }

  async onStart(sessionId: string) {
    delete this.answerExpected;
    if (!sessionId) {
      return this.onStatus();
    }
    const session = authSessionManager.getSession(sessionId);
    if (session) {
      try {
        await this.onAuthSessionMessage(session, "", true);
      } catch (err) {
        session.reportFailure(err, "telegram");
      }
    } else {
      throw new Error(`session not found with ${sessionId}`);
    }
  }

  async onPhoto(ctx: ContextMessageUpdate) {
    try {
      const b = await this.camera.takePhoto();
      await this.sendMessage("sending your file...");
      logger.debug("Photo size=%s", b ? b.length + "" : "null");
      await this.bot.telegram.sendPhoto(
        this.chatId,
        { source: b },
        { caption: moment().format("YYYYMMDDHHmmss") }
      );
    } catch (err) {
      logger.error("failed taking photo", err);
      await this.sendMessage("failed taking photos");
    }
  }

  async onAuthSessionEvent(evt: IAuthSessionEvt) {
    const session = evt.session;
    const authState = evt.session.authState;

    if (authState.isFinal()) {
      this.sessionId = undefined;
      await this.onStatus(session.lastMessage);
    } else if (
      authState === AuthStates.AUTHED_WAITING_DISARM_DURATION &&
      (!session.lastError || (evt.origin === this.name && session.lastError)) // ignore error message of other authenticators
    ) {
      await this.sendMessage(
        session.lastMessage +
          " Enter the disarm time (ex: 4h for 4 hours) or just type 0 to disable the alarm."
      );
    } else if (authState === AuthStates.STARTED && session.tries > 0 && evt.origin === this.name) {
      await this.sendMessage(session.lastMessage);
    }
  }

  async authenticate(session: IAuthSession) {
    logger.debug("Starting telegram authentication");

    this.sessionId = session.sessionId;

    session.registerListener(async (evt: IAuthSessionEvt) => {
      try {
        await this.onAuthSessionEvent(evt);
      } catch (err) {
        logger.error(err);
      }
    });
    try {
      if (!this.chatId) {
        logger.debug(
          `No chat present sending notification to ${this.telegramCfg.botName} for session ${
            session.sessionId
          }`
        );
        this.chatId = `@${this.telegramCfg.channel}`;
        const startSessionMarkup = Markup.inlineKeyboard([
          [
            Markup.urlButton(
              "Authenticate",
              `telegram.me/${this.telegramCfg.botName}?start=${session.sessionId}`
            )
          ]
        ]);
        logger.debug(`sending message to @${this.telegramCfg.channel}`);
        await this.sendMessage("", {
          reply_markup: startSessionMarkup
        });
      } else {
        await this.sendMessage(AUTH_MSG);
      }
    } catch (err) {
      session.reportFailure(err, this.name);
    }
  }

  async onAuthSessionMessage(session: IAuthSession, text: string, isStart?: boolean) {
    logger.debug("session %s", session);

    const sessionUpdatedSinceLastMessage = session.lastUpdateTime > this.lastMessageSentTime;
    if (session.authState === AuthStates.STARTED && !isStart) {
      const enteredPassword = text;
      session.authenticate(enteredPassword, this.name);
    } else if (session.authState === AuthStates.STARTED && isStart) {
      return await this.sendMessage(AUTH_MSG);
    } else if (
      session.authState === AuthStates.AUTHED_WAITING_DISARM_DURATION &&
      !sessionUpdatedSinceLastMessage // if session has been updated it means that the msg.txt is the password, ignore it it is not required anymore
    ) {
      try {
        session.setDisarmDuration(text, false, "telegram");
      } catch (err) {
        // ignore
      }
    } else {
      logger.debug("no action to be executed on session");
    }
  }

  async onDisarm(disarmDurationStr: string) {
    if (disarmDurationStr && disarmDurationStr.length > 0) {
      let duration;
      try {
        duration = util.parseDuration(disarmDurationStr, "h");
      } catch (err) {
        this.answerExpected = "disarm";
        return await this.sendMessage("invalid disarm time, enter again");
      }
      try {
        await this.alarm.disarm(duration);
        await this.onStatus();
      } catch (err) {
        logger.error("Failed disarming alarm", err);
        await this.sendMessage("disarm failure");
      }
    } else {
      this.answerExpected = "disarm";
      await this.sendMessage("How long alarm must be disarmed ?");
    }
  }

  async onCam() {
    try {
      const hasStarted = await this.camera.toggleWebCam();
      if (hasStarted) {
        return await this.sendMessage(
          "Live stream started, check out https://www.youtube.com/live_dashboard"
        );
      } else {
        return await this.sendMessage("Live stream stopped");
      }
    } catch (err) {
      logger.error("failed executing webcam operation ", err);
      await this.sendMessage("webcam command failed");
    }
  }

  async onCamStatus() {
    try {
      const status = await this.camera.getStatus();
      return await this.sendMessage("camera is " + status);
    } catch (err) {
      logger.error("failed getting camera status", err);
      await this.sendMessage("Could not get camera status, reason=" + err.message);
    }
  }
}

File: rpicalarm/agents/telegram.py
# -*- coding: utf-8 -*-
import logging
import datetime

from telegram import Bot, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Updater, CommandHandler, RegexHandler, DispatcherHandlerStop, ConversationHandler

from rpicalarm.util import getLogger
from .. import events, run_async, AuthFailureReason, AlarmState

LOGGER = getLogger(__name__)
getLogger("telegram").setLevel(logging.ERROR)

AUTH_MSG = "Intrusion detected. What is your password ?"

CONV_AUTH, CONV_SET_DISARM_TIME = range(2)


class Telegram(object):
    """
    Performs end-user interaction.
    """

    def __init__(self, alarm, camera, **cfg):
        self.bot_name = cfg["bot_name"]
        self.user_id = int(cfg["user_id"])
        self.user_name = cfg["user_name"]
        self.channel = cfg["channel"]
        self.bot = Bot(cfg["bot_token"])
        self.alarm = alarm
        self.camera = camera
        self.chat_id = None
        self.session = None
        self.conv_handler = None
        self._start()
        self._register_events_handlers()

        events.authenticator_started(self)

    def _register_events_handlers(self):
        events.authentication_succeeded += self.on_authentication_successful
        events.authentication_ended += self.on_authentication_ended
        events.authentication_failed += self.on_authentication_failed
        events.alarm_authenticating += self.on_authentication_required
        events.alarm_alarming += self._send_status
        events.alarm_disabled += self._send_status
        events.alarm_disarmed += self._send_status
        events.alarm_armed += self._send_status

    def handle_save_chat_id(self, _, update):
        from_user_id = update.message.from_user.id
        if from_user_id != self.user_id:
            LOGGER.debug("Received unauthenticated message from user %s", from_user_id)
            raise DispatcherHandlerStop("Unauthenticated message")
        self.chat_id = update.message.chat_id
        LOGGER.debug("Set Telegram chat_id %s", update.message.chat_id)

    def handle_debug(self, _, update):
        LOGGER.debug("Received Telegram bot message: %s", update.message.text)

    def handle_get_status(self, *_):
        msg = "status: {0}".format(self.alarm.state)
        if self.alarm.state == AlarmState.DISARMED:
            msg += "\nWill be re-armed in {0}".format(self.alarm.get_readable_disarm_time())
        self._send_message(msg)

    def handle_take_photo(self, *_):
        photo = None
        try:
            with self.camera.take_photo_io() as photo:
                self._send_message("sending your file...")
                self.bot.send_photo(self.chat_id, photo, timeout= 60,
                                    caption=datetime.datetime.now().strftime('%H:%M:%S %d/%m/%Y'))
        except Exception:
            LOGGER.exception("Failed taking photo")
            self._send_message("Failed taking photo")

    def handle_enable(self, *_):
        self._update_state(AlarmState.ARMED)

    def handle_disable(self, *_):
        self.session = None
        self._update_state(AlarmState.DISABLED)

    def handle_auth_update(self, _, update):
        LOGGER.debug("auth update %s", repr(update))
        session = self.session
        if session is None:
            LOGGER.debug("No session found ending conversation")
            return ConversationHandler.END

        if session.authenticate(self, update.message.text):
            LOGGER.debug("Ending auth conversation")
            return CONV_SET_DISARM_TIME
        else:
            self._send_message("Authentication failed. {0} tries remaining".format(
                session.remaining_tries))
            return None if self.alarm.state == AlarmState.AUTHENTICATING else ConversationHandler.END

    def handle_disarm_update(self, _, update):
        try:
            self.alarm.set_disarm_time(update.message.text.strip())
            return ConversationHandler.END
        except Exception:
            LOGGER.exception("Could not set disarm time")
            self._send_message("Invalid duration format")
            return None

    def handle_cam(self, *_):
        try:
            if self.camera.toggle_web_stream():
                self._send_message(
                    "Live stream started, check out https://www.youtube.com/live_dashboard")
            else:
                self._send_message("Live stream stopped")
        except Exception:
            LOGGER.exception("Failed toggling camera")
            self._send_message("Toggling camera failed")

    def handle_cam_status(self, *_):
        LOGGER.debug("Executing handle_cam_status")
        try:
            status = self.camera.get_state()
            self._send_message("Camera is {}".format(status))
        except Exception:
            LOGGER.exception("Failed getting camera status")
            self._send_message("Getting status failed")

    def _update_state(self, new_state):
        try:
            self.alarm.update_state(new_state)
        except Exception:
            LOGGER.exception("Failed updating state to %s", new_state)
            self._send_message("Failed updating state, current state is %s", self.alarm.state)

    def _get_conv_key(self):
        return (self.chat_id,)

    def _authenticate(self, session):
        try:
            if not self.chat_id:
                self.chat_id = "@{0}".format(self.channel)
                LOGGER.debug("No chat id present, sending invite to join to chat")
                mark_up = InlineKeyboardMarkup([[InlineKeyboardButton(
                    text="Authenticate", url="telegram.me/{0}?start={1}".format(self.bot_name, session.id))]])
                self._send_message("", reply_markup=mark_up)
            else:
                self.conv_handler.update_state(CONV_AUTH, self._get_conv_key())
                self._send_message(AUTH_MSG)
        except Exception:
            events.authentication_failed(self, session, AuthFailureReason.AUTHENTICATOR_FAILURE)
            LOGGER.exception("Failed sending authentication message")

    @run_async
    def _send_status(self, *_):
        if self.chat_id is not None and self.bot is not None:
            self.handle_get_status()

    @run_async
    def _send_message_async(self, *args, **kwargs):
        return self._send_message(*args, **kwargs)

    def _send_message(self, text, reply_markup=None):
        LOGGER.debug("Chat id is #%s#", self.chat_id)
        self.bot.send_message(self.chat_id,
                              "[Alarm] {0}".format(text), reply_markup=reply_markup)

    def error_callback(self, _, update, error):
        LOGGER.error("Update \"%s\" caused error \"%s\"", update, error)

    def handle_chat_start(self, _, update):
        LOGGER.debug("starting chat %s", repr(update))
        if self.session is not None:
            self._send_message_async(AUTH_MSG)
            return CONV_AUTH
        return None

    @run_async
    def on_authentication_required(self, _, session):
        self.session = session
        self._authenticate(session)

    @run_async
    def on_authentication_successful(self, *_):
        self._send_message(
            "You have been authenticated. Enter the disarm time (ex: 4h for 4 hours) or just type 0 to disable the alarm.")

    @run_async
    def on_authentication_failed(self, _origin, _session, reason):
        if reason == AuthFailureReason.TIMEOUT:
            self._send_message("Authentication timed-out")

    def on_authentication_ended(self, *_):
        self.session = None

    def _start(self):
        try:
            updater = Updater(bot=self.bot)
            dispatcher = updater.dispatcher
            dispatcher.add_handler(RegexHandler('.*', self.handle_debug), group=1)
            dispatcher.add_handler(RegexHandler('.*', self.handle_save_chat_id), group=2)
            dispatcher.add_handler(CommandHandler("status", self.handle_get_status), group=3)
            dispatcher.add_handler(CommandHandler("photo", self.handle_take_photo), group=3)
            dispatcher.add_handler(CommandHandler("disable", self.handle_disable), group=3)
            dispatcher.add_handler(CommandHandler("enable", self.handle_enable), group=3)
            dispatcher.add_handler(CommandHandler("cam", self.handle_cam), group=3)
            dispatcher.add_handler(CommandHandler("camstatus", self.handle_cam_status), group=3)

            self.conv_handler = ConversationHandler(
                entry_points=[CommandHandler('start', self.handle_chat_start)],
                states={
                    CONV_AUTH: [RegexHandler('.*', self.handle_auth_update)],
                    CONV_SET_DISARM_TIME: [RegexHandler('.*', self.handle_disarm_update)]
                },
                fallbacks=[],  # CommandHandler('cancel', cancel),
                per_user=False
            )
            dispatcher.add_handler(self.conv_handler, group=4)

            dispatcher.add_error_handler(self.error_callback)
            updater.start_polling(timeout=10)
        except Exception:
            LOGGER.exception('Telegram Updater failed to start with error')
        else:
            LOGGER.info("thread running")
